proposals/route-blinding.md

```diff
@@ -0,0 +1,242 @@
+senders arbitrarily update amounts and lock times, and reuse a blinded route multiple times (which
+is useful when retrying a failed route or using multi-part payments).
+
+The downside compared to rendezvous is that senders have more leeway to probe by changing various
+variables, so the scheme needs to explicitly defend against probing attacks and may be less private.
+
+Some use-cases where route blinding is useful include:
+
+* Recipient anonymity when receiving payments
+* Using unannounced channels in invoices without revealing them
+* Forcing a payment to go through a specific set of intermediaries that can witness the payment
+* Providing anonymous reply paths for onion messages (offers, stuckless payments, etc)
+
+### Overview
+
+At a high level, route blinding works by having the recipient choose a route to himself, and then
+blind each node and channel along the path with ECDH. The recipient then includes the blinded route
+and a secret in the invoice, which allows each node in the blinded route to incrementally unblind
+the payloads.
+
+This scheme requires all the nodes in the blinded route and the sender to activate support for the
+feature. It only becomes effective once a big enough share of the network supports it.
+
+### Notations
+
+* A node `N(i)`'s `node_id` is defined as: `P(i) = k(i) * G`.
+* Blinded `node_id`s are defined as: `B(i) = b(i) * G`.
+* Ephemeral public keys are defined as: `E(i) = e(i) * G`.
+
+### Requirements
+
+A node `N(r)` wants to provide a blinded route `N(r) <- ... <- N(1) <- N(0)` that must be used
+to receive onion messages.
+
+* The channels used along that route may be either announced or unannounced.
+* When used for payments, intermediate nodes in the blinded route MUST NOT learn `payment_secret`.
+* Intermediate nodes in the blinded route MUST NOT learn the `node_id` or `scid` of other
+  intermediate nodes except for their immediate predecessor or successor.
+* Intermediate nodes in the blinded route MUST NOT learn their distance to the recipient `N(r)`.
+* Senders MUST NOT learn the real `node_id`s and `scid`s of the blinded intermediate hops after the
+  introduction point `N(0)`.
+* If `N(r)` creates multiple blinded routes to herself, senders MUST NOT be able to tell that these
+  routes lead to the same recipient (unless this information is leaked by higher layers of the
+  protocol, such as using the same `payment_hash`).
+
+### Encrypted blob
+
+Route blinding introduces a new TLV field to the onion `tlv_payload`: the `encrypted_blob`.
+This blob is used to carry the blinded `scid` to use when forwarding the message and may be
+extended with additional data in the future. It uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`enctlv`)
+2. data:
+    * [`...*byte`:`enctlv`]
+
+Once decrypted, the content of this encrypted blob is itself a TLV stream that may contain any
+tlv record defined in Bolt 4 (onion TLV namespace).
```

Comment:
Is this mandatory than `enctlv` includes tlv record `scid` and `eph_pubkey` ?

Reply:
Only scid is mandatory, the eph_pubkey cannot be provided in the onion, because you need it to decrypt the onion (there would be a chicken-and-egg problem otherwise). I think it would be helpful if I included ascii diagrams of the packets for the full flow, what do you think?

Reply:
Yeah I'm a big fan of ascii diagrams. So `eph_pubkey` would be constructed like we're doing right now but by recipient and swap-in in onion-packet by introduction point

Reply:
Done, the ASCII flow should help!